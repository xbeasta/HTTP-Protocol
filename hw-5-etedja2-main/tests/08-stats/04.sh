#!/bin/bash

STATS=$'GET /stats HTTP/1.1\r\n\r\n'

PORT=$@

PING_REQ=$'GET /ping HTTP/1.1\r\n\r\n'
PING_RESP=$'HTTP/1.1 200 OK\r\nContent-Length: 4\r\n\r\npong'

send_ping() {

    echo -n "$PING_RESP" &> /dev/null &
}

ECHO_REQ=$'GET /echo HTTP/1.1\r
Host: localhost:9999\r
Connection: keep-alive\r
sec-ch-ua: "Not?A_Brand";v="8", "Chromium";v="108", "Google Chrome";v="108"\r
sec-ch-ua-mobile: ?0\r
sec-ch-ua-platform: "Linux"\r
DNT: 1\r
Upgrade-Insecure-Requests: 1\r
User-Agent: Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/108.0.0.0 Safari/537.36\r
Accept: text/html,application/xhtml+xml,application/xml;q=0.9,image/avif,image/webp,image/apng,*/*;q=0.8,application/signed-exchange;v=b3;q=0.9\r
Sec-Fetch-Site: none\r
Sec-Fetch-Mode: navigate\r
Sec-Fetch-User: ?1\r
Sec-Fetch-Dest: document\r
Accept-Encoding: gzip, deflate, br\r
Accept-Language: en-US,en;q=0.9,pt;q=0.8\r
Cookie: _ga=GA1.1.149590090.1676066969\r\n\r\n'

send_echo() {

    echo -n "$ECHO_REQ" | nc 127.0.0.1 $PORT &> /dev/null &

}

RANDOM_FILE_DATA=`dd if=/dev/urandom count=1000 bs=4096 2> /dev/null | base64`
echo -n "$RANDOM_DATA" > file

send_file() {
    curl -s http://127.0.0.1:$PORT/file &> /dev/null
}


send_error() {
    curl -s http://127.0.0.1:$PORT/f &> /dev/null &
}

RANDOM_DATA_POST=`dd if=/dev/urandom count=1 bs=1000 2> /dev/null | base64 | head -c 1000`
echo -n "$RANDOM_DATA_POST" > expected_post

send_write() {
    EOL=$'\r'
    REQUEST="POST /write HTTP/1.1$EOL
Content-Length: 1000$EOL
$EOL
$RANDOM_DATA_POST$EOL$EOL$EOL$EOL"

    echo -n "$REQUEST" | nc 127.0.0.1 $PORT | tail -n +4 | diff expected_post - &
}

send_read() {
    REQUEST=$'GET /read HTTP/1.1\r\n\r\n'

    echo -n "$REQUEST" | nc 127.0.0.1 $PORT | tail -n +4 | diff expected_post - &
}

send_write

wait


for j in $(seq 100); do
    for i in $(seq 8); do
        send_echo
        send_ping
        send_file
        send_error
        send_write
        send_read
    done

    wait

    echo -n "$STATS" | nc 127.0.0.1 $PORT
    echo ""
done

rm expected_* &> /dev/null
rm file_* &> /dev/null

echo -n "$STATS" | nc 127.0.0.1 $PORT
echo ""