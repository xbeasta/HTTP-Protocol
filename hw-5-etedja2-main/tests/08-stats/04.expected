HTTP/1.1 200 OK
Content-Length: 76

Requests: 33
Header bytes: 1321
Body bytes: 22456
Errors: 8
Error bytes: 208
HTTP/1.1 200 OK
Content-Length: 77

Requests: 66
Header bytes: 2640
Body bytes: 43988
Errors: 16
Error bytes: 416
HTTP/1.1 200 OK
Content-Length: 77

Requests: 99
Header bytes: 3959
Body bytes: 65521
Errors: 24
Error bytes: 624
HTTP/1.1 200 OK
Content-Length: 78

Requests: 132
Header bytes: 5278
Body bytes: 87054
Errors: 32
Error bytes: 832
HTTP/1.1 200 OK
Content-Length: 80

Requests: 165
Header bytes: 6597
Body bytes: 108588
Errors: 40
Error bytes: 1040
HTTP/1.1 200 OK
Content-Length: 80

Requests: 198
Header bytes: 7916
Body bytes: 130124
Errors: 48
Error bytes: 1248
HTTP/1.1 200 OK
Content-Length: 80

Requests: 231
Header bytes: 9235
Body bytes: 151660
Errors: 56
Error bytes: 1456
HTTP/1.1 200 OK
Content-Length: 81

Requests: 264
Header bytes: 10554
Body bytes: 173196
Errors: 64
Error bytes: 1664
HTTP/1.1 200 OK
Content-Length: 81

Requests: 297
Header bytes: 11873
Body bytes: 194733
Errors: 72
Error bytes: 1872
HTTP/1.1 200 OK
Content-Length: 81

Requests: 330
Header bytes: 13192
Body bytes: 216270
Errors: 80
Error bytes: 2080
HTTP/1.1 200 OK
Content-Length: 81

Requests: 363
Header bytes: 14511
Body bytes: 237807
Errors: 88
Error bytes: 2288
HTTP/1.1 200 OK
Content-Length: 81

Requests: 396
Header bytes: 15830
Body bytes: 259344
Errors: 96
Error bytes: 2496
HTTP/1.1 200 OK
Content-Length: 82

Requests: 429
Header bytes: 17149
Body bytes: 280881
Errors: 104
Error bytes: 2704
HTTP/1.1 200 OK
Content-Length: 82

Requests: 462
Header bytes: 18468
Body bytes: 302419
Errors: 112
Error bytes: 2912
HTTP/1.1 200 OK
Content-Length: 82

Requests: 495
Header bytes: 19787
Body bytes: 323957
Errors: 120
Error bytes: 3120
HTTP/1.1 200 OK
Content-Length: 82

Requests: 528
Header bytes: 21106
Body bytes: 345495
Errors: 128
Error bytes: 3328
HTTP/1.1 200 OK
Content-Length: 82

Requests: 561
Header bytes: 22425
Body bytes: 367033
Errors: 136
Error bytes: 3536
HTTP/1.1 200 OK
Content-Length: 82

Requests: 594
Header bytes: 23744
Body bytes: 388571
Errors: 144
Error bytes: 3744
HTTP/1.1 200 OK
Content-Length: 82

Requests: 627
Header bytes: 25063
Body bytes: 410109
Errors: 152
Error bytes: 3952
HTTP/1.1 200 OK
Content-Length: 82

Requests: 660
Header bytes: 26382
Body bytes: 431647
Errors: 160
Error bytes: 4160
HTTP/1.1 200 OK
Content-Length: 82

Requests: 693
Header bytes: 27701
Body bytes: 453185
Errors: 168
Error bytes: 4368
HTTP/1.1 200 OK
Content-Length: 82

Requests: 726
Header bytes: 29020
Body bytes: 474723
Errors: 176
Error bytes: 4576
HTTP/1.1 200 OK
Content-Length: 82

Requests: 759
Header bytes: 30339
Body bytes: 496261
Errors: 184
Error bytes: 4784
HTTP/1.1 200 OK
Content-Length: 82

Requests: 792
Header bytes: 31658
Body bytes: 517799
Errors: 192
Error bytes: 4992
HTTP/1.1 200 OK
Content-Length: 82

Requests: 825
Header bytes: 32977
Body bytes: 539337
Errors: 200
Error bytes: 5200
HTTP/1.1 200 OK
Content-Length: 82

Requests: 858
Header bytes: 34296
Body bytes: 560875
Errors: 208
Error bytes: 5408
HTTP/1.1 200 OK
Content-Length: 82

Requests: 891
Header bytes: 35615
Body bytes: 582413
Errors: 216
Error bytes: 5616
HTTP/1.1 200 OK
Content-Length: 82

Requests: 924
Header bytes: 36934
Body bytes: 603951
Errors: 224
Error bytes: 5824
HTTP/1.1 200 OK
Content-Length: 82

Requests: 957
Header bytes: 38253
Body bytes: 625489
Errors: 232
Error bytes: 6032
HTTP/1.1 200 OK
Content-Length: 82

Requests: 990
Header bytes: 39572
Body bytes: 647027
Errors: 240
Error bytes: 6240
HTTP/1.1 200 OK
Content-Length: 83

Requests: 1023
Header bytes: 40891
Body bytes: 668565
Errors: 248
Error bytes: 6448
HTTP/1.1 200 OK
Content-Length: 83

Requests: 1056
Header bytes: 42210
Body bytes: 690104
Errors: 256
Error bytes: 6656
HTTP/1.1 200 OK
Content-Length: 83

Requests: 1089
Header bytes: 43529
Body bytes: 711643
Errors: 264
Error bytes: 6864
HTTP/1.1 200 OK
Content-Length: 83

Requests: 1122
Header bytes: 44848
Body bytes: 733182
Errors: 272
Error bytes: 7072
HTTP/1.1 200 OK
Content-Length: 83

Requests: 1155
Header bytes: 46167
Body bytes: 754721
Errors: 280
Error bytes: 7280
HTTP/1.1 200 OK
Content-Length: 83

Requests: 1188
Header bytes: 47486
Body bytes: 776260
Errors: 288
Error bytes: 7488
HTTP/1.1 200 OK
Content-Length: 83

Requests: 1221
Header bytes: 48805
Body bytes: 797799
Errors: 296
Error bytes: 7696
HTTP/1.1 200 OK
Content-Length: 83

Requests: 1254
Header bytes: 50124
Body bytes: 819338
Errors: 304
Error bytes: 7904
HTTP/1.1 200 OK
Content-Length: 83

Requests: 1287
Header bytes: 51443
Body bytes: 840877
Errors: 312
Error bytes: 8112
HTTP/1.1 200 OK
Content-Length: 83

Requests: 1320
Header bytes: 52762
Body bytes: 862416
Errors: 320
Error bytes: 8320
HTTP/1.1 200 OK
Content-Length: 83

Requests: 1353
Header bytes: 54081
Body bytes: 883955
Errors: 328
Error bytes: 8528
HTTP/1.1 200 OK
Content-Length: 83

Requests: 1386
Header bytes: 55400
Body bytes: 905494
Errors: 336
Error bytes: 8736
HTTP/1.1 200 OK
Content-Length: 83

Requests: 1419
Header bytes: 56719
Body bytes: 927033
Errors: 344
Error bytes: 8944
HTTP/1.1 200 OK
Content-Length: 83

Requests: 1452
Header bytes: 58038
Body bytes: 948572
Errors: 352
Error bytes: 9152
HTTP/1.1 200 OK
Content-Length: 83

Requests: 1485
Header bytes: 59357
Body bytes: 970111
Errors: 360
Error bytes: 9360
HTTP/1.1 200 OK
Content-Length: 83

Requests: 1518
Header bytes: 60676
Body bytes: 991650
Errors: 368
Error bytes: 9568
HTTP/1.1 200 OK
Content-Length: 84

Requests: 1551
Header bytes: 61995
Body bytes: 1013189
Errors: 376
Error bytes: 9776
HTTP/1.1 200 OK
Content-Length: 84

Requests: 1584
Header bytes: 63314
Body bytes: 1034729
Errors: 384
Error bytes: 9984
HTTP/1.1 200 OK
Content-Length: 85

Requests: 1617
Header bytes: 64633
Body bytes: 1056269
Errors: 392
Error bytes: 10192
HTTP/1.1 200 OK
Content-Length: 85

Requests: 1650
Header bytes: 65952
Body bytes: 1077810
Errors: 400
Error bytes: 10400
HTTP/1.1 200 OK
Content-Length: 85

Requests: 1683
Header bytes: 67271
Body bytes: 1099351
Errors: 408
Error bytes: 10608
HTTP/1.1 200 OK
Content-Length: 85

Requests: 1716
Header bytes: 68590
Body bytes: 1120892
Errors: 416
Error bytes: 10816
HTTP/1.1 200 OK
Content-Length: 85

Requests: 1749
Header bytes: 69909
Body bytes: 1142433
Errors: 424
Error bytes: 11024
HTTP/1.1 200 OK
Content-Length: 85

Requests: 1782
Header bytes: 71228
Body bytes: 1163974
Errors: 432
Error bytes: 11232
HTTP/1.1 200 OK
Content-Length: 85

Requests: 1815
Header bytes: 72547
Body bytes: 1185515
Errors: 440
Error bytes: 11440
HTTP/1.1 200 OK
Content-Length: 85

Requests: 1848
Header bytes: 73866
Body bytes: 1207056
Errors: 448
Error bytes: 11648
HTTP/1.1 200 OK
Content-Length: 85

Requests: 1881
Header bytes: 75185
Body bytes: 1228597
Errors: 456
Error bytes: 11856
HTTP/1.1 200 OK
Content-Length: 85

Requests: 1914
Header bytes: 76504
Body bytes: 1250138
Errors: 464
Error bytes: 12064
HTTP/1.1 200 OK
Content-Length: 85

Requests: 1947
Header bytes: 77823
Body bytes: 1271679
Errors: 472
Error bytes: 12272
HTTP/1.1 200 OK
Content-Length: 85

Requests: 1980
Header bytes: 79142
Body bytes: 1293220
Errors: 480
Error bytes: 12480
HTTP/1.1 200 OK
Content-Length: 85

Requests: 2013
Header bytes: 80461
Body bytes: 1314761
Errors: 488
Error bytes: 12688
HTTP/1.1 200 OK
Content-Length: 85

Requests: 2046
Header bytes: 81780
Body bytes: 1336302
Errors: 496
Error bytes: 12896
HTTP/1.1 200 OK
Content-Length: 85

Requests: 2079
Header bytes: 83099
Body bytes: 1357843
Errors: 504
Error bytes: 13104
HTTP/1.1 200 OK
Content-Length: 85

Requests: 2112
Header bytes: 84418
Body bytes: 1379384
Errors: 512
Error bytes: 13312
HTTP/1.1 200 OK
Content-Length: 85

Requests: 2145
Header bytes: 85737
Body bytes: 1400925
Errors: 520
Error bytes: 13520
HTTP/1.1 200 OK
Content-Length: 85

Requests: 2178
Header bytes: 87056
Body bytes: 1422466
Errors: 528
Error bytes: 13728
HTTP/1.1 200 OK
Content-Length: 85

Requests: 2211
Header bytes: 88375
Body bytes: 1444007
Errors: 536
Error bytes: 13936
HTTP/1.1 200 OK
Content-Length: 85

Requests: 2244
Header bytes: 89694
Body bytes: 1465548
Errors: 544
Error bytes: 14144
HTTP/1.1 200 OK
Content-Length: 85

Requests: 2277
Header bytes: 91013
Body bytes: 1487089
Errors: 552
Error bytes: 14352
HTTP/1.1 200 OK
Content-Length: 85

Requests: 2310
Header bytes: 92332
Body bytes: 1508630
Errors: 560
Error bytes: 14560
HTTP/1.1 200 OK
Content-Length: 85

Requests: 2343
Header bytes: 93651
Body bytes: 1530171
Errors: 568
Error bytes: 14768
HTTP/1.1 200 OK
Content-Length: 85

Requests: 2376
Header bytes: 94970
Body bytes: 1551712
Errors: 576
Error bytes: 14976
HTTP/1.1 200 OK
Content-Length: 85

Requests: 2409
Header bytes: 96289
Body bytes: 1573253
Errors: 584
Error bytes: 15184
HTTP/1.1 200 OK
Content-Length: 85

Requests: 2442
Header bytes: 97608
Body bytes: 1594794
Errors: 592
Error bytes: 15392
HTTP/1.1 200 OK
Content-Length: 85

Requests: 2475
Header bytes: 98927
Body bytes: 1616335
Errors: 600
Error bytes: 15600
HTTP/1.1 200 OK
Content-Length: 86

Requests: 2508
Header bytes: 100246
Body bytes: 1637876
Errors: 608
Error bytes: 15808
HTTP/1.1 200 OK
Content-Length: 86

Requests: 2541
Header bytes: 101565
Body bytes: 1659418
Errors: 616
Error bytes: 16016
HTTP/1.1 200 OK
Content-Length: 86

Requests: 2574
Header bytes: 102884
Body bytes: 1680960
Errors: 624
Error bytes: 16224
HTTP/1.1 200 OK
Content-Length: 86

Requests: 2607
Header bytes: 104203
Body bytes: 1702502
Errors: 632
Error bytes: 16432
HTTP/1.1 200 OK
Content-Length: 86

Requests: 2640
Header bytes: 105522
Body bytes: 1724044
Errors: 640
Error bytes: 16640
HTTP/1.1 200 OK
Content-Length: 86

Requests: 2673
Header bytes: 106841
Body bytes: 1745586
Errors: 648
Error bytes: 16848
HTTP/1.1 200 OK
Content-Length: 86

Requests: 2706
Header bytes: 108160
Body bytes: 1767128
Errors: 656
Error bytes: 17056
HTTP/1.1 200 OK
Content-Length: 86

Requests: 2739
Header bytes: 109479
Body bytes: 1788670
Errors: 664
Error bytes: 17264
HTTP/1.1 200 OK
Content-Length: 86

Requests: 2772
Header bytes: 110798
Body bytes: 1810212
Errors: 672
Error bytes: 17472
HTTP/1.1 200 OK
Content-Length: 86

Requests: 2805
Header bytes: 112117
Body bytes: 1831754
Errors: 680
Error bytes: 17680
HTTP/1.1 200 OK
Content-Length: 86

Requests: 2838
Header bytes: 113436
Body bytes: 1853296
Errors: 688
Error bytes: 17888
HTTP/1.1 200 OK
Content-Length: 86

Requests: 2871
Header bytes: 114755
Body bytes: 1874838
Errors: 696
Error bytes: 18096
HTTP/1.1 200 OK
Content-Length: 86

Requests: 2904
Header bytes: 116074
Body bytes: 1896380
Errors: 704
Error bytes: 18304
HTTP/1.1 200 OK
Content-Length: 86

Requests: 2937
Header bytes: 117393
Body bytes: 1917922
Errors: 712
Error bytes: 18512
HTTP/1.1 200 OK
Content-Length: 86

Requests: 2970
Header bytes: 118712
Body bytes: 1939464
Errors: 720
Error bytes: 18720
HTTP/1.1 200 OK
Content-Length: 86

Requests: 3003
Header bytes: 120031
Body bytes: 1961006
Errors: 728
Error bytes: 18928
HTTP/1.1 200 OK
Content-Length: 86

Requests: 3036
Header bytes: 121350
Body bytes: 1982548
Errors: 736
Error bytes: 19136
HTTP/1.1 200 OK
Content-Length: 86

Requests: 3069
Header bytes: 122669
Body bytes: 2004090
Errors: 744
Error bytes: 19344
HTTP/1.1 200 OK
Content-Length: 86

Requests: 3102
Header bytes: 123988
Body bytes: 2025632
Errors: 752
Error bytes: 19552
HTTP/1.1 200 OK
Content-Length: 86

Requests: 3135
Header bytes: 125307
Body bytes: 2047174
Errors: 760
Error bytes: 19760
HTTP/1.1 200 OK
Content-Length: 86

Requests: 3168
Header bytes: 126626
Body bytes: 2068716
Errors: 768
Error bytes: 19968
HTTP/1.1 200 OK
Content-Length: 86

Requests: 3201
Header bytes: 127945
Body bytes: 2090258
Errors: 776
Error bytes: 20176
HTTP/1.1 200 OK
Content-Length: 86

Requests: 3234
Header bytes: 129264
Body bytes: 2111800
Errors: 784
Error bytes: 20384
HTTP/1.1 200 OK
Content-Length: 86

Requests: 3267
Header bytes: 130583
Body bytes: 2133342
Errors: 792
Error bytes: 20592
HTTP/1.1 200 OK
Content-Length: 86

Requests: 3300
Header bytes: 131902
Body bytes: 2154884
Errors: 800
Error bytes: 20800
HTTP/1.1 200 OK
Content-Length: 86

Requests: 3301
Header bytes: 131941
Body bytes: 2154970
Errors: 800
Error bytes: 20800
